cdisc_rules_engine/utilities/base_report.py

```diff
@@ -0,0 +1,167 @@
+            "Dataset",
+            "USUBJID",
+            "Record",
+            "Sequence",
+            "Variable(s)",
+            "Value(s)"
+        ]
+        """
+        errors = []
+        for result in validation_result.results or []:
+            if result.get("errors", []) and result.get("executionStatus") == "success":
+                variables = result.get("variables", [])
+                for error in result.get("errors"):
+                    error_item = {
+                        "rule_id": validation_result.id,
+                        "message": result.get("message"),
+                        "severity": validation_result.severity,
+                        "domain": result.get("domain"),
+                        "uSubjId": error.get("uSubjId", ""),
+                        "row": error.get("row", ""),
+                        "seq": error.get("seq", ""),
+                    }
+
+                    if self._item_type == "list":
+                        error_item["variables"] = ", ".join(variables)
+                        error_item["values"] = ", ".join(
+                            [
+                                str(error.get("value", {}).get(variable))
+                                for variable in variables
+                            ]
+                        )
+                        errors = errors + [[*error_item.values()]]
+                    elif self._item_type == "dict":
+                        error_item["variables"] = variables
+                        error_item["values"] = [
+                            str(error.get("value", {}).get(variable))
+                            for variable in variables
+                        ]
+                        errors = errors + [error_item]
+        return errors
+
+    def get_rules_report_data(self) -> List[List]:
+        """
+        Generates the rules report data that goes into the excel export.
+        Each row is represented by a list or a dict containing the following information:
+        [
+            "RuleID",
+            "Version",
+            "Message",
+            "Status"
+        ]
+        """
+        rules_report = []
+        for validation_result in self._results:
+            rules_item = {
+                "rule_id": validation_result.id,
+                "version": "1",
+                "message": validation_result.message,
+                "status": "SUCCESS"
+                if validation_result.execution_status == "success"
+                else "SKIPPED",
+            }
+            if self._item_type == "list":
+                rules_report.append([*rules_item.values()])
+            elif self._item_type == "dict":
+                rules_report.append(rules_item)
+
+        return sorted(
+            rules_report,
+            key=lambda x: x[0] if (self._item_type == "list") else x["rule_id"],
+        )
+
+    @abstractmethod
+    def get_export() -> Union[dict, Workbook]:
+        raise NotImplementedError
```

Comment:
`abstractmethod` decorator and raising a `NotImplementedError` are 2 different ways to do the same thing but they should not be combined. 

The difference is that raising a `NotImplementedError` is the old style whereas `abstractmethod` decorator is the new style. A method marked with `abstractmethod` decorator must not contain any implementation (raising an error is also an implementation), the method can only contain `pass` keyword or a docstring. 

In terms of our project, we tend to use the new style but some examples of raising a `NotImplementedError` still exist in the project and my suggestion would be to unify the style later, not in this PR.

Reply:
Aleksei for the explanation, will update it. I think that flake8 will catch it as well, so if we have it, that might help to avoid such issues.

Reply:
Updated